compiler/rustc_mir/src/transform/large_enums.rs

```diff
@@ -0,0 +1,214 @@
+use crate::transform::MirPass;
+use crate::util::patch::MirPatch;
+use rustc_middle::mir::*;
+use rustc_middle::ty::{self, Const, List, ParamEnv, TyCtxt};
+use rustc_target::abi::{Size, Variants};
+
+/// A pass that seeks to optimize unnecessary moves of large enum types, if there is a large
+/// enough discrepanc between them
+pub struct EnumSizeOpt<const DISCREPANCY: u64>;
```

Comment:
pondering: is it really important for `D` to be compile-time here?  Can we save monomorphization work by just making it be `pub struct EnumSizeOpt { discrepancy: u64 }`?